File: backend/worker/clipEmbed.py
#!/usr/bin/env python3
"""
Whole-image embedding extractor — the CHARACTER counterpart to faceEmbed.py.

A person is judged by their face (insightface / ArcFace). A CHARACTER — a
personified mango, a mascot, an animated fruit — has no face to detect, so its
identity has to ride on the whole frame. This process embeds the whole image
with CLIP and reports the vector; faceQc.js decides whether it is the same
character, exactly as it does for a face, because CLIP vectors compare by angle
too.

Same wire protocol as faceEmbed.py so the Node side is the same class pointed at
a different script: one JSON request per line on stdin, one JSON result per line
on stdout, diagnostics to stderr, a `ready` line before any work so the model
load never sits inside a job's lease.

    {"id": "job-42-1", "path": "/tmp/still.png", "expect_aspect": 0.8}
    → {"id": "job-42-1", "ok": true, "embedding": [...512 floats...],
       "width": 1024, "height": 1280, "aspect": 0.8, "aspect_ok": true}

Unlike the face extractor there is NO detection step and no `faces` count: a
whole-image embedding always exists for any image that opens, so a character
frame is never rejected for "no face" — the identity judgement is the cosine
alone. The face-only structural rejects (no_face, multiple_faces, hands) are
skipped for characters on the judgement side too (faceQc, subjectType).

Model: openai/clip-vit-base-patch32 — 512-d image features, the same width as
ArcFace by coincidence, not by design; the two vector spaces are unrelated and a
character's reference mean must be built from CLIP vectors, never mixed with a
face mean.

Setup (shares the torch env with indicf5_tts.py):
    pip install torch transformers pillow
"""

import json
import sys


MODEL_NAME = "openai/clip-vit-base-patch32"
EMBED_DIM = 512


def log(message):
    """Diagnostics go to stderr — stdout carries only JSON results."""
    print(message, file=sys.stderr, flush=True)


def load_model():
    try:
        import torch
        from transformers import CLIPModel, CLIPProcessor
    except ImportError:
        log("torch/transformers not installed — pip install torch transformers pillow")
        raise

    # CPU is fine: one 224x224 forward pass is milliseconds, and this avoids
    # fighting ComfyUI/IndicF5 for unified memory on the Mac.
    model = CLIPModel.from_pretrained(MODEL_NAME)
    model.eval()
    processor = CLIPProcessor.from_pretrained(MODEL_NAME)
    return {"torch": torch, "model": model, "processor": processor}


def analyse(ctx, path, expect_aspect=None, aspect_tolerance=0.02):
    from PIL import Image

    torch = ctx["torch"]
    model = ctx["model"]
    processor = ctx["processor"]

    with Image.open(path) as img:
        img = img.convert("RGB")
        width, height = img.size
        inputs = processor(images=img, return_tensors="pt")

    with torch.no_grad():
        feats = model.get_image_features(**inputs)

    # L2-normalise so the stored vector is a direction. cosineSimilarity would
    # normalise anyway, but a unit vector is what the mean-of-a-seed-set wants:
    # averaging raw-magnitude CLIP features lets a brighter frame weigh more.
    feats = feats / feats.norm(p=2, dim=-1, keepdim=True)
    embedding = [float(v) for v in feats[0].tolist()]

    result = {
        "ok": True,
        "embedding": embedding,
        "width": width,
        "height": height,
    }

    if expect_aspect is not None:
        actual = width / height if height else 0
        result["aspect"] = round(actual, 4)
        result["aspect_ok"] = abs(actual - expect_aspect) <= aspect_tolerance

    return result


def main():
    ctx = load_model()
    log(f"clipEmbed ready — {MODEL_NAME}")
    # The worker waits for this line before sending work, so it never pays the
    # model load time inside a job's lease.
    print(json.dumps({"ready": True, "model": MODEL_NAME, "dim": EMBED_DIM}), flush=True)

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        request_id = None
        try:
            request = json.loads(line)
            request_id = request.get("id")
            out = analyse(
                ctx,
                request["path"],
                expect_aspect=request.get("expect_aspect"),
                aspect_tolerance=request.get("aspect_tolerance", 0.02),
            )
            out["id"] = request_id
        except Exception as err:  # noqa: BLE001 — one bad image must not kill the process
            out = {"id": request_id, "ok": False, "error": f"{type(err).__name__}: {err}"}

        try:
            payload = json.dumps(out)
        except Exception as err:  # noqa: BLE001
            payload = json.dumps({
                "id": request_id,
                "ok": False,
                "error": f"result was not serialisable: {type(err).__name__}: {err}",
            })

        print(payload, flush=True)


def probe():
    """
    Can this machine embed whole images?

    Imports the dependencies and exits — it does NOT load the model, which pulls
    ~600MB the first time and several hundred MB into memory after. The API asks
    this once at boot to decide whether to claim character `qc`/`embed` jobs
    itself or leave them for a worker on a machine that can.
    """
    try:
        import torch  # noqa: F401
        import transformers  # noqa: F401
        print(json.dumps({"ok": True, "model": MODEL_NAME}), flush=True)
        return 0
    except Exception as err:  # noqa: BLE001
        print(json.dumps({"ok": False, "error": f"{type(err).__name__}: {err}"}), flush=True)
        return 1


if __name__ == "__main__":
    if "--probe" in sys.argv:
        sys.exit(probe())
    main()

File: backend/worker/clipEmbed.js
'use strict';

/**
 * Node side of the whole-image (CLIP) embedder — the CHARACTER counterpart to
 * faceEmbed.js.
 *
 * Same shape as FaceEmbedder, deliberately: one long-lived Python process,
 * line-delimited JSON, requests correlated by id out of a pending map so a slow
 * image can never hand its result to the wrong caller. The only differences are
 * the script it drives (clipEmbed.py) and that a CLIP result has no `faces`
 * count — a whole-image embedding always exists, so there is no "no face" case
 * to report. Identity is the cosine alone; faceQc's character branch skips the
 * face-only structural rejects.
 *
 * Kept as its own class rather than parameterising FaceEmbedder because the two
 * point at different models with different load costs and different env knobs,
 * and a shared class that switched scripts by a flag would be one object whose
 * behaviour depended on how it was constructed — exactly the ambiguity the
 * embedder factory exists to make explicit instead.
 */

const { spawn } = require('child_process');
const path = require('path');
const readline = require('readline');

class ClipEmbedError extends Error {
  constructor(message, { permanent = false } = {}) {
    super(message);
    this.name = 'ClipEmbedError';
    this.permanent = permanent;
  }
}

class ClipEmbedder {
  constructor({
    // Falls back to the face extractor's interpreter: on the Mac both models
    // live in the same torch env, so one CLIP_EMBED_PYTHON override is rarely
    // needed and FACE_EMBED_PYTHON is the sensible default.
    python = process.env.CLIP_EMBED_PYTHON || process.env.FACE_EMBED_PYTHON || 'python3',
    script = path.join(__dirname, 'clipEmbed.py'),
    startupTimeoutMs = 180_000,   // CLIP pulls ~600MB the first time it runs
    requestTimeoutMs = 60_000,
  } = {}) {
    this.python = python;
    this.script = script;
    this.startupTimeoutMs = startupTimeoutMs;
    this.requestTimeoutMs = requestTimeoutMs;

    this.proc = null;
    this.ready = null;
    this.pending = new Map();
    this.seq = 0;
    this.dim = null;
  }

  /** Idempotent — several callers may await start() and get the same process. */
  start() {
    if (this.ready) return this.ready;

    this.ready = new Promise((resolve, reject) => {
      const proc = spawn(this.python, [this.script], { stdio: ['pipe', 'pipe', 'pipe'] });
      this.proc = proc;

      const timer = setTimeout(() => {
        reject(new ClipEmbedError(`clipEmbed.py did not become ready within ${this.startupTimeoutMs}ms`));
        this.stop();
      }, this.startupTimeoutMs);

      const rl = readline.createInterface({ input: proc.stdout });
      rl.on('line', (line) => {
        let message;
        try {
          message = JSON.parse(line);
        } catch {
          return; // Not our protocol. Ignore rather than crash the worker.
        }

        if (message.ready) {
          clearTimeout(timer);
          this.dim = message.dim;
          resolve(this);
          return;
        }

        const entry = this.pending.get(message.id);
        if (!entry) return;
        this.pending.delete(message.id);
        clearTimeout(entry.timer);
        entry.resolve(message);
      });

      proc.stderr.on('data', (chunk) => process.stderr.write(`[clipEmbed] ${chunk}`));

      proc.on('exit', (code) => {
        clearTimeout(timer);
        const err = new ClipEmbedError(`clipEmbed.py exited with code ${code}`);
        for (const [, entry] of this.pending) {
          clearTimeout(entry.timer);
          entry.reject(err);
        }
        this.pending.clear();
        this.ready = null;
        this.proc = null;
        reject(err);
      });

      proc.on('error', (err) => {
        clearTimeout(timer);
        reject(new ClipEmbedError(`Cannot start ${this.python}: ${err.message}`, { permanent: true }));
      });
    });

    return this.ready;
  }

  async isReady() {
    try {
      await this.start();
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Measure one image. Returns what was measured, never a verdict.
   *
   * The result shape matches faceEmbed's where the two overlap (`ok`,
   * `embedding`, `width`, `height`, `aspect`, `aspect_ok`) so qcStage reads them
   * the same way; there is simply no `faces`/`significant_faces` — for a
   * character those are meaningless, and qcStage's `?? 0` fallback lands on a
   * count the character branch of faceQc ignores.
   */
  async embed(imagePath, { expectAspect = null, aspectTolerance = 0.02 } = {}) {
    await this.start();

    const id = `r${++this.seq}`;
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new ClipEmbedError(`Embedding ${imagePath} timed out after ${this.requestTimeoutMs}ms`));
      }, this.requestTimeoutMs);

      this.pending.set(id, { resolve, reject, timer });

      this.proc.stdin.write(
        `${JSON.stringify({ id, path: imagePath, expect_aspect: expectAspect, aspect_tolerance: aspectTolerance })}\n`
      );
    }).then((result) => {
      if (!result.ok) {
        throw new ClipEmbedError(`Embedding failed for ${imagePath}: ${result.error}`, { permanent: true });
      }
      return result;
    });
  }

  async embedMany(paths, opts = {}) {
    const out = [];
    for (const p of paths) out.push(await this.embed(p, opts));
    return out;
  }

  stop() {
    if (this.proc) {
      this.proc.stdin.end();
      this.proc.kill();
      this.proc = null;
    }
    this.ready = null;
  }
}

module.exports = { ClipEmbedder, ClipEmbedError };
